### PR TITLE
Auto create threads for suggestions

### LINKED ISSUE #468: Auto creation of threads in suggestion channel
Right now the bot adds up and down reaction on every message also on messages where someone created a thread. Ex:-
image

**Describe the solution you'd like**
Let the bot create a thread for each suggestion and rest discussion will go inside it

**Describe alternatives you've considered**
Analyse the message if it is an thread created message then don't add reactions to it

### PR DESCRIPTION
**Pull request by Tanish-Azad**
Auto create thread for suggestions in suggestions channel.

Closes #468 .

image

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: Have you tested this locally in your test server already? This appears to be wrong. If I read it right, it says "if someone created a thread, lets create another thread".

**Review by Zabuzard — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: Nvm. I read it wrong. I would still be interested in how it looks in practice. Can you post a screenshot?

Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: when the bot creates a thread it does not sends a thread creation message
https://drive.google.com/file/d/1-JGBDCSbo2IYxgn-PKY8szrBlFTbMrcZ/view?usp=sharing
the testing bot have admin perms

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: Thats good to know, but could you please send a screenshot of this piece of code, your current implementation, running live? It looks like you just created another mini project to showcase it, instead of actually trying out the real code with our real project.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: Also, the code is failing the spotless check right now.

Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:52]: image
This is from config file:
```json
"suggestions": {
    "channelPattern": "suggestions",
    "upVoteEmoteName": "thumbsup",
    "downVoteEmoteName": "thumbsdown"
  }
```

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:50]: This `if` should not only affect thread creation but also the emoji-reactions. All in all, it should basically affect the whole method. Please move it up into the general checks at the top, i.e. the initial `if` that `return`s.

**Comment by Tanish-Azad** [comment]
what to do about the spotless check?

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:51]: Please put it into a method, such as `createDiscussionThread(...);`

Ideally the code would then look like:
```java
if (...) {
  return;
}

...

createDiscussionThread(...);
reactWith(...);
reactWith(...);
```

**Review by Zabuzard — CHANGES_REQUESTED** [review]
Also, the code is not passing spotless.

**Comment by Zabuzard** [comment]
> what to do about the spotless check?

If you have set up the code locally and use the gradle commands, it would run spotless already for you. See our wiki for details on that.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:50]: ok 👍🏻

Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:50]:
```java
        if ((event.getAuthor().isBot() || event.isWebhookMessage() || !event.isFromGuild()) && !event.getChannel().getType().isThread()) {
            return;
        }
```
like this?

Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:50]: got it
```java
        if (event.getAuthor().isBot() || event.isWebhookMessage() || !event.isFromGuild() || event.getChannel().getType().isThread()) {
            return;
        }
```

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java:51]: done

**Comment by Tanish-Azad** [comment]
what the thread title should be "Discussion on User's suggestion" or "Discussion for User's suggestion" or something else

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: Better, but still not exactly what I was looking for. You are still checking whether the message is a thread-message. But instead, you should just look for "not ordinary message". I.e. the code should only react on ordinary text messages, not on any other fancy type. Right now you only exclude thread-types but what about all other unwanted types.

**Comment by Zabuzard** [comment]
> what the thread title should be "Discussion on User's suggestion" or "Discussion for User's suggestion" or something else

I dont care. Imo it is fine as u did it :)

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: i don't see any other type of message is suggestions channel there is only the actual suggestions and thread creation messages, we already handled the thread creation message. What is left now?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: See, your logic is just the reverse. You do not actually want to check "is not a thread channel" but instead "is a message/guild/text channel" or whatever that channel is. I.e. something that exclusively identifies the channel you are interested in.

https://ci.dv8tion.net/job/JDA5/javadoc/net/dv8tion/jda/api/entities/ChannelType.html

Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: > is a message/guild/text channel

it is a message because MessageReceivedEvent only happens when it is a message
it is a guild channel because we already have the condition to check that inside the if statement
it is a text channel because messages can only be sent in a text channel + we know that suggestions channel is going to be a text channel

so now all we need to check is that it is not a thread inside the suggestions channel

hope that makes sense

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: Sure that this can even happen? Due to the name-pattern it only listens to messages in this particular channel matching that name anyways.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: I just tried it out. It seems that the condition is obsolete. Thread messages do not enter the `onMessageReceived` due to the channel name pattern.

**Review by Zabuzard — CHANGES_REQUESTED** [review]
This is a bit more tricky, but currently the UX suffers from the generic thread names:

issue

I think we can not release it like that and need some way to distinguish them.

I would suggest we determine the title based on the first x characters from the message. It is not perfect, but enough to make it clear what the thread is about. So in this case `Discussion for 'I have a great idea'` or just `I have a great idea` without the fuzz around it.

Code-wise, it could look something like `message.getContent()` and then `substring(0, Math.min(MAX_LENGTH, content.length())`. Where the max length could be something around 60 maybe.

If you want, you can also make the substring smarter by cutting off at the last "word" to not have something like `I have a gre` where its cut in the middle of the word. The code for something like already exists in `ImplicitAskListener#createTitle` where you can get some inspiration, if you want to.

File: application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java
Comment (Tanish-Azad) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: hmm yes that can be the case i didn't though of that, so should i remove the if condition and test it again?

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/basic/SuggestionsUpDownVoter.java]: yeah, thanks

**Comment by Zabuzard** [comment]
> image is this better?

Yes, but you have to ensure that its within the limit of thread names. So limit it to 60 characters for example. And if you do not want to cut it off in between a word, use a logic as shown earlier.

**Comment by Tanish-Azad** [comment]
yes, i did
```java
    private static void createThread(@NotNull Message message) {
        String messageContent = message.getContentRaw();

        String title = messageContent;
        if (messageContent.length() >= TITLE_MAX_LENGTH) {
            int lastWordEnd = messageContent.lastIndexOf(' ', TITLE_MAX_LENGTH);

            if (lastWordEnd == -1) {
                lastWordEnd = TITLE_MAX_LENGTH;
            }

            title = messageContent.substring(0, lastWordEnd);
        }

        message
            .createThreadChannel(title)
            .queue();
    }
```

**Comment by Tanish-Azad** [comment]
image

**Comment by Tanish-Azad** [comment]
Any thread created will automatically invite the author and the member who created the thread

**Review by Zabuzard — APPROVED** [review]
Seems to work well

tests

**Comment by Zabuzard** [comment]
Waiting for a second approval. Otherwise merging in latest 7 days.

**Review by Heatmanofurioso — APPROVED** [review]

**Pull request merged**
